[Serialization] Reloading simulation from file doesn't work anymore

**Describe the bug**
It looks like saving a simulation into a file is working (though hard to check, but at least a file is created) but when trying to restore the simulation from a file an error is raised

**To Reproduce**
Steps to reproduce the behavior:
1. Go to `Create Simulation From File.gaml`
2. Run the first experiment up to cycle 5
3. now try to run the second simulation
4. See error:
```
1 occurence in 2. Reload Simulation0 at cycle 0: Java error: I/O error
in Simulation 0
in agent 2. Reload Simulation0
IOException: java.lang.NullPointerException
org.nustaq.serialization.FSTObjectInput.readObject(FSTObjectInput.java:377)
org.nustaq.serialization.FSTConfiguration.asObject(FSTConfiguration.java:801)
gama.extension.serialize.binary.BinarySerialiser.restoreAgentFromBytes(BinarySerialiser.java:95)
gama.extension.serialize.binary.BinarySerialisation.restoreFromBytes(BinarySerialisation.java:179)
gama.extension.serialize.binary.BinarySerialisation.restoreFromFile(BinarySerialisation.java:131)
gama.extension.serialize.gaml.CreateAgentsFromSerialisedFileDelegate.createAgents(CreateAgentsFromSerialisedFileDelegate.java:51)
gama.gaml.statements.CreateStatement.privateExecuteIn(CreateStatement.java:460)
in create simulation from: input ;
```
And here is what I have in the eclipse console:
```
unable to decode:
java.io.IOException: java.lang.NullPointerException: Cannot load from object array because the return value of "org.nustaq.serialization.FSTClazzInfo$FSTFieldInfo.getPossibleClasses()" is null
	at org.nustaq.serialization.FSTObjectInput.readObject(FSTObjectInput.java:377)
	at org.nustaq.serialization.FSTConfiguration.asObject(FSTConfiguration.java:801)
	at gama.extension.serialize.binary.BinarySerialiser.restoreAgentFromBytes(BinarySerialiser.java:95)
	at gama.extension.serialize.binary.BinarySerialisation.restoreFromBytes(BinarySerialisation.java:179)
	at gama.extension.serialize.binary.BinarySerialisation.restoreFromFile(BinarySerialisation.java:131)
	at gama.extension.serialize.gaml.CreateAgentsFromSerialisedFileDelegate.createAgents(CreateAgentsFromSerialisedFileDelegate.java:51)
	at gama.gaml.statements.CreateStatement.privateExecuteIn(CreateStatement.java:460)
	at gama.gaml.statements.CreateStatement.privateExecuteIn(CreateStatement.java:1)
	at gama.gaml.statements.AbstractStatement.executeOn(AbstractStatement.java:47)
	at gama.gaml.statements.AbstractStatementSequence.executeOn(AbstractStatementSequence.java:59)
	at gama.core.runtime.ExecutionScope.execute(ExecutionScope.java:509)
	at gama.core.runtime.IScope.execute(IScope.java:443)
	at gama.core.runtime.IScope.execute(IScope.java:414)
	at gama.gaml.statements.AbstractStatementSequence.privateExecuteIn(AbstractStatementSequence.java:69)
	at gama.gaml.statements.AbstractStatementSequenceWithArgs.privateExecuteIn(AbstractStatementSequenceWithArgs.java:59)
	at gama.gaml.statements.AbstractStatement.executeOn(AbstractStatement.java:47)
	at gama.gaml.statements.AbstractStatementSequence.executeOn(AbstractStatementSequence.java:59)
	at gama.core.runtime.ExecutionScope.execute(ExecutionScope.java:509)
	at gama.core.runtime.IScope.execute(IScope.java:443)
	at gama.core.metamodel.agent.MinimalAgent.init(MinimalAgent.java:243)
	at gama.core.kernel.experiment.ExperimentAgent.init(ExperimentAgent.java:456)
	at gama.core.runtime.ExecutionScope.init(ExecutionScope.java:601)
	at gama.core.kernel.experiment.DefaultExperimentController.schedule(DefaultExperimentController.java:200)
	at gama.core.kernel.experiment.ExperimentAgent.schedule(ExperimentAgent.java:536)
	at gama.core.kernel.experiment.ExperimentPlan.open(ExperimentPlan.java:713)
	at gama.core.kernel.experiment.ExperimentPlan.open(ExperimentPlan.java:771)
	at gama.core.kernel.experiment.DefaultExperimentController.processUserCommand(DefaultExperimentController.java:72)
	at gama.core.kernel.experiment.AbstractExperimentController.lambda$0(AbstractExperimentController.java:63)
	at java.base/java.lang.Thread.run(Thread.java:1570)
Caused by: java.lang.NullPointerException: Cannot load from object array because the return value of "org.nustaq.serialization.FSTClazzInfo$FSTFieldInfo.getPossibleClasses()" is null
	at org.nustaq.serialization.FSTObjectInput.readObjectWithHeader(FSTObjectInput.java:523)
	at org.nustaq.serialization.FSTObjectInput.readObjectInternal(FSTObjectInput.java:487)
	at org.nustaq.serialization.FSTObjectInput.readObject(FSTObjectInput.java:447)
	at org.nustaq.serialization.FSTObjectInput.readObject(FSTObjectInput.java:375)
	... 28 more
```


**Expected behavior**
Restoring simulations works again

**Additional context**
May be related to previous serialization issues: #244 and #246

--- a/gama.extension.network/src/gama/extension/network/common/CompositeGamaMessage.java
+++ b/gama.extension.network/src/gama/extension/network/common/CompositeGamaMessage.java
@@ -1,16 +1,15 @@
 /*******************************************************************************************************
  *
- * CompositeGamaMessage.java, in gama.extension.network, is part of the source code of the
- * GAMA modeling and simulation platform (v.2024-06).
+ * CompositeGamaMessage.java, in gama.extension.network, is part of the source code of the GAMA modeling and simulation
+ * platform (v.2024-06).
  *
  * (c) 2007-2024 UMI 209 UMMISCO IRD/SU & Partners (IRIT, MIAT, ESPACE-DEV, CTU)
  *
  * Visit https://github.com/gama-platform/gama for license information and contacts.
- * 
+ *
  ********************************************************************************************************/
 package gama.extension.network.common;
 
-import gama.core.common.interfaces.ISerialisationConstants;
 import gama.core.messaging.GamaMessage;
 import gama.core.runtime.IScope;
 import gama.extension.serialize.binary.BinarySerialisation;
@@ -33,8 +32,7 @@ public class CompositeGamaMessage extends GamaMessage {
 	 */
 	public CompositeGamaMessage(final IScope scope, final GamaMessage message) {
 		super(scope, message.getSender(), message.getReceivers(), message.getContents(scope));
-		this.contents = BinarySerialisation.saveToString(scope, message.getContents(scope),
-				ISerialisationConstants.BINARY_FORMAT, true);
+		this.contents = BinarySerialisation.saveToString(scope, message.getContents(scope));
 		// this.contents = StreamConverter.convertNetworkObjectToStream(scope, message.getContents(scope));
 		this.emissionTimeStamp = message.getEmissionTimestamp();
 		this.setUnread(true);

--- a/gama.extension.network/src/gama/extension/network/common/Connector.java
+++ b/gama.extension.network/src/gama/extension/network/common/Connector.java
@@ -1,12 +1,12 @@
 /*******************************************************************************************************
  *
- * Connector.java, in gama.extension.network, is part of the source code of the
- * GAMA modeling and simulation platform (v.2024-06).
+ * Connector.java, in gama.extension.network, is part of the source code of the GAMA modeling and simulation platform
+ * (v.2024-06).
  *
  * (c) 2007-2024 UMI 209 UMMISCO IRD/SU & Partners (IRIT, MIAT, ESPACE-DEV, CTU)
  *
  * Visit https://github.com/gama-platform/gama for license information and contacts.
- * 
+ *
  ********************************************************************************************************/
 package gama.extension.network.common;
 
@@ -17,7 +17,6 @@ import java.util.LinkedList;
 import java.util.List;
 import java.util.Map;
 
-import gama.core.common.interfaces.ISerialisationConstants;
 import gama.core.messaging.GamaMessage;
 import gama.core.metamodel.agent.IAgent;
 import gama.core.runtime.IScope;
@@ -189,11 +188,9 @@ public abstract class Connector implements IConnector {
 				if (cmsg.getSender() instanceof IAgent) {
 					cmsg.setSender(sender.getAttribute(INetworkSkill.NET_AGENT_NAME));
 				}
-				final NetworkMessage msg = MessageFactory
-						.buildNetworkMessage((String) sender.getAttribute(INetworkSkill.NET_AGENT_NAME), receiver,
-								BinarySerialisation.saveToString(sender.getScope(), cmsg,
-										ISerialisationConstants.BINARY_FORMAT,
-										true) /* StreamConverter.convertObjectToStream(sender.getScope(), cmsg) */);
+				final NetworkMessage msg =
+						MessageFactory.buildNetworkMessage((String) sender.getAttribute(INetworkSkill.NET_AGENT_NAME),
+								receiver, BinarySerialisation.saveToString(sender.getScope(), cmsg));
 				this.sendMessage(sender, receiver, MessageFactory.packMessage(msg));
 			} else {
 				this.sendMessage(sender, receiver, content.getContents(sender.getScope()).toString());

--- a/gama.extension.serialize/models/Serialization/models/Create Simulation From File.gaml
+++ b/gama.extension.serialize/models/Serialization/models/Create Simulation From File.gaml
@@ -13,14 +13,13 @@ import "Base Model.gaml"
 experiment "1. Save Simulation" type: gui parent: Base {
 	
 	
-	string format <- "binary";
 	string file_path <- "../includes/saved_simulation.simulation";
 	
 	text "Run the simulation until cycle 5, when it will be saved in a file and quit" font: font("Helvetica", 14, #bold);
 	parameter "File path" var: file_path ;
 	
 	reflex store when: cycle = 5 {		
-		save simulation to: file_path format: format;
+		save simulation to: file_path ;
 		do die;	
 	}	
 
@@ -48,7 +47,7 @@ experiment "3. Restore Simulation" type: gui parent: Base{
 	parameter "File to read" var: input <- file("../includes/saved_simulation.simulation");
 	
 	// We "restore" the simulation from the file. As it happens each time the simulation reaches 10 cycles, it loops forever between 5 and 10 cycles. 
-	reflex when: cycle=10 {
+	reflex when: simulation.cycle=10 {
 		restore simulation from: input;	
 	}
 

--- a/gama.extension.serialize/src/gama/extension/serialize/binary/BinarySerialisation.java
+++ b/gama.extension.serialize/src/gama/extension/serialize/binary/BinarySerialisation.java
@@ -10,23 +10,27 @@
  ********************************************************************************************************/
 package gama.extension.serialize.binary;
 
+import static gama.core.common.util.FileUtils.constructAbsoluteFilePath;
+import static gama.core.util.ByteArrayZipper.unzip;
+import static gama.core.util.ByteArrayZipper.zip;
+import static java.nio.file.Files.readAllBytes;
+import static java.nio.file.StandardOpenOption.CREATE;
+import static java.nio.file.StandardOpenOption.TRUNCATE_EXISTING;
+import static java.nio.file.StandardOpenOption.WRITE;
+
 import java.io.File;
 import java.io.IOException;
 import java.io.OutputStream;
-import java.nio.ByteBuffer;
 import java.nio.file.Files;
 import java.nio.file.Path;
-import java.nio.file.StandardOpenOption;
 
 import gama.core.common.interfaces.ISerialisationConstants;
-import gama.core.common.util.FileUtils;
 import gama.core.kernel.simulation.SimulationAgent;
 import gama.core.metamodel.agent.IAgent;
 import gama.core.metamodel.agent.SerialisedAgent;
 import gama.core.runtime.GAMA;
 import gama.core.runtime.IScope;
 import gama.core.runtime.exceptions.GamaRuntimeException;
-import gama.core.util.ByteArrayZipper;
 
 /**
  * The Class BinarySerialisationReader.
@@ -52,22 +56,10 @@ public class BinarySerialisation implements ISerialisationConstants {
 	 */
 	public static Object createFromFile(final IScope scope, final String path) {
 		try {
-			byte[] all = Files.readAllBytes(Path.of(FileUtils.constructAbsoluteFilePath(scope, path, true)));
-			return createFromBytes(scope, all);
+			return createFromBytes(scope, readAllBytes(Path.of(constructAbsoluteFilePath(scope, path, true))));
 		} catch (IOException e) {
 			throw GamaRuntimeException.create(e, scope);
 		}
-	}
-
-	/**
-	 * Checks if is serialisation header.
-	 *
-	 * @param b
-	 *            the b
-	 * @return true, if is serialisation header
-	 */
-	private static boolean isSerialisationHeader(final byte b) {
-		return b == GAMA_OBJECT_IDENTIFIER || b == GAMA_AGENT_IDENTIFIER;
 	}
 
 	/**
@@ -84,8 +76,7 @@ public class BinarySerialisation implements ISerialisationConstants {
 	public static Object createFromString(final IScope scope, final String string) {
 		if (string == null || string.isBlank()) return null;
 		try {
-			byte[] all = string.getBytes(ISerialisationConstants.STRING_BYTE_ARRAY_CHARSET);
-			return createFromBytes(scope, all);
+			return createFromBytes(scope, string.getBytes(STRING_BYTE_ARRAY_CHARSET));
 		} catch (Throwable e) {
 			e.printStackTrace();
 			try {
@@ -108,13 +99,7 @@ public class BinarySerialisation implements ISerialisationConstants {
 	 * @date 31 oct. 2023
 	 */
 	public static Object createFromBytes(final IScope scope, final byte[] bytes) {
-		byte type = bytes[0];
-		if (!isSerialisationHeader(type)) throw GamaRuntimeException.error("Not a GAMA serialisation record", scope);
-		boolean zip = bytes[1] == COMPRESSED;
-		byte[] some = removeIdentifiers(bytes);
-		if (zip) { some = ByteArrayZipper.unzip(some); }
-		return type == GAMA_OBJECT_IDENTIFIER ? PROCESSOR.createObjectFromBytes(scope, some)
-				: PROCESSOR.createAgentFromBytes(scope, some);
+		return PROCESSOR.createObjectFromBytes(scope, unzip(bytes));
 	}
 
 	/**
@@ -127,8 +112,7 @@ public class BinarySerialisation implements ISerialisationConstants {
 	 */
 	public static void restoreFromFile(final IAgent agent, final String path) {
 		try {
-			byte[] all = Files.readAllBytes(Path.of(path));
-			if (isSerialisationHeader(all[0])) { restoreFromBytes(agent, all); }
+			restoreFromBytes(agent, Files.readAllBytes(Path.of(path)));
 		} catch (IOException e) {
 			throw GamaRuntimeException.create(e, agent.getScope());
 		}
@@ -146,11 +130,8 @@ public class BinarySerialisation implements ISerialisationConstants {
 	 */
 	public static void restoreFromString(final IAgent agent, final String string) {
 		try {
-			byte[] all = string.getBytes(ISerialisationConstants.STRING_BYTE_ARRAY_CHARSET);
-			restoreFromBytes(agent, all);
+			restoreFromBytes(agent, string.getBytes(STRING_BYTE_ARRAY_CHARSET));
 		} catch (Throwable e) {
-			// e.printStackTrace();
-			// The string is maybe a path ?
 			try {
 				restoreFromFile(agent, string);
 			} catch (Throwable ex) {
@@ -172,11 +153,7 @@ public class BinarySerialisation implements ISerialisationConstants {
 	 * @date 8 août 2023
 	 */
 	public static void restoreFromBytes(final IAgent sim, final byte[] bytes) throws IOException {
-		if (bytes[0] != GAMA_AGENT_IDENTIFIER) throw new IOException("Not an agent serialisation record");
-		boolean zip = bytes[1] == COMPRESSED;
-		byte[] some = removeIdentifiers(bytes);
-		if (zip) { some = ByteArrayZipper.unzip(some); }
-		PROCESSOR.restoreAgentFromBytes(sim, some);
+		PROCESSOR.restoreAgentFromBytes(sim, unzip(bytes));
 	}
 
 	/**
@@ -197,14 +174,13 @@ public class BinarySerialisation implements ISerialisationConstants {
 	 *            whether to include the "history" of the agent in the serialisation. Only applicables to simulations
 	 * @date 31 oct. 2023
 	 */
-	public static final void saveToFile(final IScope scope, final Object o, final String path, final String format,
-			final boolean zip, final boolean includingHistory) {
-		try (OutputStream os = Files.newOutputStream(new File(path).toPath(), StandardOpenOption.APPEND,
-				StandardOpenOption.CREATE, StandardOpenOption.WRITE)) {
+	public static final void saveToFile(final IScope scope, final Object o, final String path,
+			final boolean includingHistory) {
+		try (OutputStream os = Files.newOutputStream(new File(path).toPath(), CREATE, WRITE, TRUNCATE_EXISTING)) {
 			if (o instanceof SimulationAgent sim) {
 				sim.setAttribute(SerialisedAgent.SERIALISE_HISTORY, includingHistory);
 			}
-			os.write(saveToBytes(scope, o, zip));
+			os.write(saveToBytes(scope, o));
 			if (o instanceof SimulationAgent sim) { sim.setAttribute(SerialisedAgent.SERIALISE_HISTORY, false); }
 		} catch (IOException e) {
 			throw GamaRuntimeException.create(e, scope);
@@ -218,9 +194,8 @@ public class BinarySerialisation implements ISerialisationConstants {
 	 * @return
 	 * @date 8 août 2023
 	 */
-	public static final String saveToString(final IScope scope, final Object sim, final String format,
-			final boolean zip) {
-		return new String(saveToBytes(scope, sim, zip), STRING_BYTE_ARRAY_CHARSET);
+	public static final String saveToString(final IScope scope, final Object sim) {
+		return new String(saveToBytes(scope, sim), STRING_BYTE_ARRAY_CHARSET);
 	}
 
 	/**
@@ -232,46 +207,8 @@ public class BinarySerialisation implements ISerialisationConstants {
 	 * @return the string
 	 * @date 21 août 2023
 	 */
-	public static final byte[] saveToBytes(final IScope scope, final Object object, final boolean zip) {
-		byte[] toSave = object instanceof IAgent a ? PROCESSOR.saveAgentToBytes(scope, a)
-				: PROCESSOR.saveObjectToBytes(scope, object);
-		if (zip) { toSave = ByteArrayZipper.zip(toSave); }
-		return addIdentifiers(toSave, object instanceof IAgent ? GAMA_AGENT_IDENTIFIER : GAMA_OBJECT_IDENTIFIER,
-				zip ? COMPRESSED : UNCOMPRESSED);
-	}
-
-	/**
-	 * Adds the byte.
-	 *
-	 * @param array
-	 *            the array
-	 * @param identifier
-	 *            the identifier
-	 * @param zip
-	 *            the zip
-	 * @return the byte[]
-	 */
-	private static byte[] addIdentifiers(final byte[] array, final byte identifier, final byte zip) {
-		if (array == null) return new byte[] { identifier, zip };
-		ByteBuffer buffer = ByteBuffer.allocate(array.length + 2);
-		buffer.put(identifier);
-		buffer.put(zip);
-		buffer.put(array);
-		return buffer.array();
-	}
-
-	/**
-	 * Removes the identifiers.
-	 *
-	 * @param array
-	 *            the tableau
-	 * @return the byte[]
-	 */
-	public static byte[] removeIdentifiers(final byte[] array) {
-		if (array == null || array.length < 2) return new byte[0];
-		byte[] newArray = new byte[array.length - 2];
-		System.arraycopy(array, 2, newArray, 0, newArray.length);
-		return newArray;
+	public static final byte[] saveToBytes(final IScope scope, final Object object) {
+		return zip(PROCESSOR.saveObjectToBytes(scope, object));
 	}
 
 	/**
@@ -285,8 +222,8 @@ public class BinarySerialisation implements ISerialisationConstants {
 	 * @return the byte[]
 	 * @date 29 déc. 2023
 	 */
-	public static final byte[] saveToBytes(final Object object, final boolean zip) {
-		return saveToBytes(GAMA.getRuntimeScope(), object, zip);
+	public static final byte[] saveToBytes(final Object object) {
+		return saveToBytes(GAMA.getRuntimeScope(), object);
 	}
 
 }

--- a/gama.extension.serialize/src/gama/extension/serialize/binary/BinarySerialisation.java
+++ b/gama.extension.serialize/src/gama/extension/serialize/binary/BinarySerialisation.java
@@ -11,8 +11,6 @@
 package gama.extension.serialize.binary;
 
 import static gama.core.common.util.FileUtils.constructAbsoluteFilePath;
-import static gama.core.util.ByteArrayZipper.unzip;
-import static gama.core.util.ByteArrayZipper.zip;
 import static java.nio.file.Files.readAllBytes;
 import static java.nio.file.StandardOpenOption.CREATE;
 import static java.nio.file.StandardOpenOption.TRUNCATE_EXISTING;
@@ -23,6 +21,8 @@ import java.io.IOException;
 import java.io.OutputStream;
 import java.nio.file.Files;
 import java.nio.file.Path;
+
+import org.nustaq.serialization.FSTConfiguration;
 
 import gama.core.common.interfaces.ISerialisationConstants;
 import gama.core.kernel.simulation.SimulationAgent;
@@ -39,6 +39,8 @@ import gama.core.runtime.exceptions.GamaRuntimeException;
  * @date 31 oct. 2023
  */
 public class BinarySerialisation implements ISerialisationConstants {
+
+	private static FSTConfiguration FST = FSTConfiguration.createDefaultConfiguration();
 
 	/** The processor. */
 	private static BinarySerialiser PROCESSOR = new BinarySerialiser();
@@ -99,7 +101,7 @@ public class BinarySerialisation implements ISerialisationConstants {
 	 * @date 31 oct. 2023
 	 */
 	public static Object createFromBytes(final IScope scope, final byte[] bytes) {
-		return PROCESSOR.createObjectFromBytes(scope, unzip(bytes));
+		return PROCESSOR.createObjectFromBytes(scope, bytes);
 	}
 
 	/**
@@ -153,7 +155,7 @@ public class BinarySerialisation implements ISerialisationConstants {
 	 * @date 8 août 2023
 	 */
 	public static void restoreFromBytes(final IAgent sim, final byte[] bytes) throws IOException {
-		PROCESSOR.restoreAgentFromBytes(sim, unzip(bytes));
+		PROCESSOR.restoreAgentFromBytes(sim, bytes);
 	}
 
 	/**
@@ -208,7 +210,7 @@ public class BinarySerialisation implements ISerialisationConstants {
 	 * @date 21 août 2023
 	 */
 	public static final byte[] saveToBytes(final IScope scope, final Object object) {
-		return zip(PROCESSOR.saveObjectToBytes(scope, object));
+		return PROCESSOR.saveObjectToBytes(scope, object);
 	}
 
 	/**

--- a/gama.extension.serialize/src/gama/extension/serialize/binary/BinarySerialiser.java
+++ b/gama.extension.serialize/src/gama/extension/serialize/binary/BinarySerialiser.java
@@ -102,19 +102,6 @@ public class BinarySerialiser implements ISerialisationConstants {
 	}
 
 	/**
-	 * Save simulation to bytes.
-	 *
-	 * @author Alexis Drogoul (alexis.drogoul@ird.fr)
-	 * @param sim
-	 *            the sim
-	 * @return the byte[]
-	 * @date 8 août 2023
-	 */
-	public byte[] saveAgentToBytes(final IScope newScope, final IAgent sim) {
-		return saveObjectToBytes(newScope, SerialisedAgent.of(sim, true));
-	}
-
-	/**
 	 * Save object to bytes.
 	 *
 	 * @param newScope
@@ -125,7 +112,7 @@ public class BinarySerialiser implements ISerialisationConstants {
 	 */
 	public byte[] saveObjectToBytes(final IScope newScope, final Object obj) {
 		inAgent = false;
-		return fst.asByteArray(obj);
+		return fst.asByteArray(obj instanceof IAgent a ? SerialisedAgent.of(a, true) : obj);
 	}
 
 	/**
@@ -142,30 +129,9 @@ public class BinarySerialiser implements ISerialisationConstants {
 	public Object createObjectFromBytes(final IScope newScope, final byte[] input) {
 		try {
 			scope = newScope;
-			return fst.asObject(input);
-		} catch (Exception e) {
-			throw GamaRuntimeException.create(e, scope);
-		} finally {
-			scope = null;
-		}
-
-	}
-
-	/**
-	 * Creates the agent from bytes.
-	 *
-	 * @param newScope
-	 *            the new scope
-	 * @param input
-	 *            the input
-	 * @return the i agent
-	 */
-	public IAgent createAgentFromBytes(final IScope newScope, final byte[] input) {
-		try {
-			scope = newScope;
 			Object o = fst.asObject(input);
 			if (o instanceof SerialisedAgent sa) return sa.recreateIn(scope);
-			return null;
+			return o;
 		} catch (Exception e) {
 			throw GamaRuntimeException.create(e, scope);
 		} finally {

--- a/gama.extension.serialize/src/gama/extension/serialize/binary/SimulationHistory.java
+++ b/gama.extension.serialize/src/gama/extension/serialize/binary/SimulationHistory.java
@@ -14,7 +14,6 @@ import java.util.LinkedList;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
 
-import gama.core.util.ByteArrayZipper;
 import gama.dev.DEBUG;
 import gama.extension.serialize.binary.SimulationHistory.SimulationHistoryNode;
 
@@ -36,7 +35,7 @@ public class SimulationHistory extends LinkedList<SimulationHistoryNode> {
 	final ExecutorService executor = Executors.newCachedThreadPool();
 
 	public void push(final byte[] state, final int cycle) {
-		executor.execute(() -> { push(new SimulationHistoryNode(ByteArrayZipper.zip(state), cycle)); });
+		executor.execute(() -> { push(new SimulationHistoryNode(state, cycle)); });
 	}
 
 }

--- a/gama.extension.serialize/src/gama/extension/serialize/binary/SimulationSerialiser.java
+++ b/gama.extension.serialize/src/gama/extension/serialize/binary/SimulationSerialiser.java
@@ -10,6 +10,8 @@
  ********************************************************************************************************/
 package gama.extension.serialize.binary;
 
+import static gama.core.util.ByteArrayZipper.zip;
+
 import java.util.LinkedList;
 
 import gama.core.common.interfaces.ISerialisationConstants;
@@ -46,7 +48,7 @@ public class SimulationSerialiser implements ISimulationRecorder, ISerialisation
 	@Override
 	public void record(final SimulationAgent sim) {
 		try {
-			byte[] state = processor.saveAgentToBytes(sim.getScope(), sim);
+			byte[] state = zip(processor.saveObjectToBytes(sim.getScope(), sim));
 			SimulationHistory history = getSimulationHistory(sim);
 			history.push(state, sim.getClock().getCycle());
 		} catch (Throwable e) {

--- a/gama.extension.serialize/src/gama/extension/serialize/binary/SimulationSerialiser.java
+++ b/gama.extension.serialize/src/gama/extension/serialize/binary/SimulationSerialiser.java
@@ -10,15 +10,12 @@
  ********************************************************************************************************/
 package gama.extension.serialize.binary;
 
-import static gama.core.util.ByteArrayZipper.zip;
-
 import java.util.LinkedList;
 
 import gama.core.common.interfaces.ISerialisationConstants;
 import gama.core.kernel.experiment.ISimulationRecorder;
 import gama.core.kernel.simulation.SimulationAgent;
 import gama.core.metamodel.agent.SerialisedAgent;
-import gama.core.util.ByteArrayZipper;
 import gama.dev.DEBUG;
 import gama.extension.serialize.binary.SimulationHistory.SimulationHistoryNode;
 
@@ -48,7 +45,7 @@ public class SimulationSerialiser implements ISimulationRecorder, ISerialisation
 	@Override
 	public void record(final SimulationAgent sim) {
 		try {
-			byte[] state = zip(processor.saveObjectToBytes(sim.getScope(), sim));
+			byte[] state = processor.saveObjectToBytes(sim.getScope(), sim);
 			SimulationHistory history = getSimulationHistory(sim);
 			history.push(state, sim.getClock().getCycle());
 		} catch (Throwable e) {
@@ -91,7 +88,7 @@ public class SimulationSerialiser implements ISimulationRecorder, ISerialisation
 				if (node != null && node.cycle() == sim.getClock().getCycle()) { node = history.pop(); }
 				if (node != null) {
 					// long startTime = System.nanoTime();
-					processor.restoreAgentFromBytes(sim, ByteArrayZipper.unzip(node.bytes()));
+					processor.restoreAgentFromBytes(sim, node.bytes());
 					// DEBUG.OUT("Deserialised in " + TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startTime) +
 					// "ms");
 				}

--- a/gama.extension.serialize/src/gama/extension/serialize/gaml/SerialisationOperators.java
+++ b/gama.extension.serialize/src/gama/extension/serialize/gaml/SerialisationOperators.java
@@ -1,7 +1,7 @@
 /*******************************************************************************************************
  *
- * SerialisationOperators.java, in gama.serialize, is part of the source code of the GAMA modeling and
- * simulation platform .
+ * SerialisationOperators.java, in gama.serialize, is part of the source code of the GAMA modeling and simulation
+ * platform .
  *
  * (c) 2007-2024 UMI 209 UMMISCO IRD/SU & Partners (IRIT, MIAT, TLU, CTU)
  *
@@ -13,22 +13,19 @@ package gama.extension.serialize.gaml;
 import java.io.ByteArrayOutputStream;
 import java.nio.charset.StandardCharsets;
 
-import gama.annotations.precompiler.IConcept;
-import gama.annotations.precompiler.IOperatorCategory;
-import gama.annotations.precompiler.ITypeProvider;
 import gama.annotations.precompiler.GamlAnnotations.doc;
 import gama.annotations.precompiler.GamlAnnotations.example;
 import gama.annotations.precompiler.GamlAnnotations.no_test;
 import gama.annotations.precompiler.GamlAnnotations.operator;
 import gama.annotations.precompiler.GamlAnnotations.test;
-import gama.annotations.precompiler.GamlAnnotations.usage;
-import gama.core.common.interfaces.IKeyword;
+import gama.annotations.precompiler.IConcept;
+import gama.annotations.precompiler.IOperatorCategory;
+import gama.annotations.precompiler.ITypeProvider;
 import gama.core.common.util.StringUtils;
 import gama.core.metamodel.agent.IAgent;
 import gama.core.runtime.GAMA;
 import gama.core.runtime.IScope;
 import gama.core.runtime.exceptions.GamaRuntimeException;
-import gama.core.util.IContainer;
 import gama.core.util.file.json.Json;
 import gama.core.util.file.json.WriterConfig;
 import gama.dev.DEBUG;
@@ -47,7 +44,6 @@ public class SerialisationOperators {
 	static {
 		DEBUG.OFF();
 	}
-
 
 	/**
 	 * To gaml.
@@ -213,7 +209,7 @@ public class SerialisationOperators {
 			see = { "to_json", "to_gaml" })
 	@no_test ()
 	public static String serialize(final IScope scope, final Object obj) {
-		return BinarySerialisation.saveToString(scope, obj, "binary", true);
+		return BinarySerialisation.saveToString(scope, obj);
 	}
 
 	/**
@@ -298,8 +294,8 @@ public class SerialisationOperators {
 			final IExpression e = GAML.getExpressionFactory().createExpr(gaml, d);
 			return scope.evaluate(e, agent).getValue();
 		} catch (final GamaRuntimeException e) {
-			GAMA.reportAndThrowIfNeeded(scope, GamaRuntimeException.error("Error in evaluating Gaml code : '" + gaml + "' in "
-					+ scope.getAgent() + Strings.LN + "Reason: " + e.getMessage(), scope), false);
+			GAMA.reportAndThrowIfNeeded(scope, GamaRuntimeException.error("Error in evaluating Gaml code : '" + gaml
+					+ "' in " + scope.getAgent() + Strings.LN + "Reason: " + e.getMessage(), scope), false);
 
 			return null;
 		}

--- a/gama.extension.serialize/src/gama/extension/serialize/gaml/SimulationSaveDelegate.java
+++ b/gama.extension.serialize/src/gama/extension/serialize/gaml/SimulationSaveDelegate.java
@@ -1,7 +1,7 @@
 /*******************************************************************************************************
  *
- * SimulationSaveDelegate.java, in gama.serialize, is part of the source code of the GAMA modeling and
- * simulation platform .
+ * SimulationSaveDelegate.java, in gama.serialize, is part of the source code of the GAMA modeling and simulation
+ * platform .
  *
  * (c) 2007-2024 UMI 209 UMMISCO IRD/SU & Partners (IRIT, MIAT, TLU, CTU)
  *
@@ -35,11 +35,10 @@ import gama.gaml.types.Types;
 public class SimulationSaveDelegate implements ISaveDelegate, ISerialisationConstants {
 
 	@Override
-	public void save(final IScope scope, final IExpression item, final File file, final SaveOptions options) throws IOException {
+	public void save(final IScope scope, final IExpression item, final File file, final SaveOptions options)
+			throws IOException {
 		Object toSave = item.value(scope);
-		if (toSave instanceof IAgent sa) {
-			BinarySerialisation.saveToFile(scope, sa, file.getPath(), options.type, true, true);
-		}
+		if (toSave instanceof IAgent sa) { BinarySerialisation.saveToFile(scope, sa, file.getPath(), true); }
 	}
 
 	@Override

--- a/gama.extension.serialize/src/org/nustaq/serialization/FSTConfiguration.java
+++ b/gama.extension.serialize/src/org/nustaq/serialization/FSTConfiguration.java
@@ -11,7 +11,6 @@
 package org.nustaq.serialization;
 
 import java.io.IOException;
-import java.io.InputStream;
 import java.io.OutputStream;
 import java.lang.ref.SoftReference;
 import java.math.BigDecimal;
@@ -55,10 +54,10 @@ import org.nustaq.serialization.serializers.FSTStringSerializer;
 import org.nustaq.serialization.serializers.FSTThrowableSerializer;
 import org.nustaq.serialization.serializers.FSTTimestampSerializer;
 import org.nustaq.serialization.util.DefaultFSTInt2ObjectMapFactory;
-import org.nustaq.serialization.util.FSTInputStream;
 import org.nustaq.serialization.util.FSTInt2ObjectMapFactory;
 import org.nustaq.serialization.util.FSTUtil;
 
+import gama.core.util.ByteArrayZipper;
 import gama.gaml.compilation.kernel.GamaClassLoader;
 
 /**
@@ -70,9 +69,8 @@ import gama.gaml.compilation.kernel.GamaClassLoader;
  */
 public class FSTConfiguration {
 
-
 	/** The stream coder factory. */
-	private StreamCoderFactory streamCoderFactory = new FSTDefaultStreamCoderFactory(this);
+	private final StreamCoderFactory streamCoderFactory = new FSTDefaultStreamCoderFactory(this);
 
 	/** The name. */
 	private String name;
@@ -216,7 +214,7 @@ public class FSTConfiguration {
 	 *            the int to object map factory
 	 * @date 30 sept. 2023
 	 */
-	public void registerIntToObjectMapFactory(final FSTInt2ObjectMapFactory intToObjectMapFactory) {
+	void registerIntToObjectMapFactory(final FSTInt2ObjectMapFactory intToObjectMapFactory) {
 		if (intToObjectMapFactory != null) { this.intToObjectMapFactory = intToObjectMapFactory; }
 	}
 
@@ -242,19 +240,8 @@ public class FSTConfiguration {
 	 * @return the instantiator
 	 * @date 30 sept. 2023
 	 */
-	public FSTClassInstantiator getInstantiator(final Class<?> clazz) {
+	FSTClassInstantiator getInstantiator(final Class<?> clazz) {
 		return instantiator;
-	}
-
-	/**
-	 * special configuration used internally for struct emulation
-	 *
-	 * @return
-	 */
-	public static FSTConfiguration createStructConfiguration() {
-		FSTConfiguration conf = new FSTConfiguration();
-		conf.setStructMode(true);
-		return conf;
 	}
 
 	/**
@@ -265,37 +252,8 @@ public class FSTConfiguration {
 	 *            the shared field infos
 	 * @date 30 sept. 2023
 	 */
-	protected FSTConfiguration() {
+	FSTConfiguration() {
 		this.fieldInfoCache = new ConcurrentHashMap<>();
-	}
-
-	/**
-	 * Gets the stream coder factory.
-	 *
-	 * @author Alexis Drogoul (alexis.drogoul@ird.fr)
-	 * @return the stream coder factory
-	 * @date 30 sept. 2023
-	 */
-	public StreamCoderFactory getStreamCoderFactory() { return streamCoderFactory; }
-
-	/**
-	 * allows to use subclassed stream codecs. Can also be used to change class loading behaviour, as clasForName is
-	 * part of a codec's interface.
-	 *
-	 * e.g. new StreamCoderFactory() {
-	 *
-	 * @Override public FSTEncoder createStreamEncoder() { return new FSTStreamEncoder(FSTConfiguration.this); }
-	 *
-	 * @Override public FSTDecoder createStreamDecoder() { return new FSTStreamDecoder(FSTConfiguration.this) { public
-	 *           Class classForName(String name) { ... } } ; } };
-	 *
-	 *           You need to work with thread locals most probably as the factory is ~global (assigned to
-	 *           fstconfiguration shared amongst streams)
-	 *
-	 * @param streamCoderFactory
-	 */
-	public void setStreamCoderFactory(final StreamCoderFactory streamCoderFactory) {
-		this.streamCoderFactory = streamCoderFactory;
 	}
 
 	/**
@@ -319,16 +277,6 @@ public class FSTConfiguration {
 		} finally {
 			cacheLock.set(false);
 		}
-	}
-
-	/**
-	 * patch default serializer lookup. set to null to delete. Should be set prior to any serialization going on
-	 * (serializer lookup is cached).
-	 *
-	 * @param del
-	 */
-	public void setSerializerRegistryDelegate(final FSTSerializerRegistryDelegate del) {
-		serializationInfoRegistry.setSerializerRegistryDelegate(del);
 	}
 
 	/** The cache lock. */
@@ -370,30 +318,6 @@ public class FSTConfiguration {
 	 * @date 30 sept. 2023
 	 */
 	public boolean isForceSerializable() { return true; }
-
-	/**
-	 * clear global deduplication caches. Useful for class reloading scenarios, else counter productive as
-	 * j.reflect.Fiwld + Construtors will be instantiated more than once per class.
-	 */
-	public static void clearGlobalCaches() {
-		FSTClazzInfo.sharedFieldSets.clear();
-		FSTDefaultClassInstantiator.constructorMap.clear();
-	}
-
-	/**
-	 * clear cached softref's and ThreadLocal.
-	 */
-	public void clearCaches() {
-		try {
-			FSTInputStream.cachedBuffer.set(null);
-			while (!cacheLock.compareAndSet(false, true)) {
-				// empty
-			}
-			cachedObjects.clear();
-		} finally {
-			cacheLock.set(false);
-		}
-	}
 
 	/**
 	 * Checks if is share references.
@@ -535,33 +459,6 @@ public class FSTConfiguration {
 	}
 
 	/**
-	 * utility for thread safety and reuse. Do not close the resulting stream. However you should close the given
-	 * InputStream 'in'
-	 *
-	 * @param in
-	 * @return
-	 */
-	public FSTObjectInput getObjectInput(final InputStream in) {
-		FSTObjectInput fstObjectInput = getIn();
-		try {
-			fstObjectInput.resetForReuse(in);
-			return fstObjectInput;
-		} catch (IOException e) {
-			FSTUtil.<RuntimeException> rethrow(e);
-		}
-		return null;
-	}
-
-	/**
-	 * Gets the object input.
-	 *
-	 * @author Alexis Drogoul (alexis.drogoul@ird.fr)
-	 * @return the object input
-	 * @date 30 sept. 2023
-	 */
-	public FSTObjectInput getObjectInput() { return getObjectInput((InputStream) null); }
-
-	/**
 	 * take the given array as input. the array is NOT copied.
 	 *
 	 * WARNING: the input streams takes over ownership and might overwrite content of this array in subsequent IO
@@ -571,6 +468,7 @@ public class FSTConfiguration {
 	 * @return
 	 */
 	public FSTObjectInput getObjectInput(final byte arr[]) {
+
 		return getObjectInput(arr, arr.length);
 	}
 
@@ -588,24 +486,6 @@ public class FSTConfiguration {
 		FSTObjectInput fstObjectInput = getIn();
 		try {
 			fstObjectInput.resetForReuseUseArray(arr, len);
-			return fstObjectInput;
-		} catch (IOException e) {
-			FSTUtil.<RuntimeException> rethrow(e);
-		}
-		return null;
-	}
-
-	/**
-	 * take the given array and copy it to input. the array IS copied
-	 *
-	 * @param arr
-	 * @param len
-	 * @return
-	 */
-	public FSTObjectInput getObjectInputCopyFrom(final byte arr[], final int off, final int len) {
-		FSTObjectInput fstObjectInput = getIn();
-		try {
-			fstObjectInput.resetForReuseCopyArray(arr, off, len);
 			return fstObjectInput;
 		} catch (IOException e) {
 			FSTUtil.<RuntimeException> rethrow(e);
@@ -670,21 +550,6 @@ public class FSTConfiguration {
 	 * @return a recycled outputstream reusing its last recently used byte[] buffer
 	 */
 	public FSTObjectOutput getObjectOutput() { return getObjectOutput((OutputStream) null); }
-
-	/**
-	 * Gets the object output.
-	 *
-	 * @author Alexis Drogoul (alexis.drogoul@ird.fr)
-	 * @param outByte
-	 *            the out byte
-	 * @return the object output
-	 * @date 30 sept. 2023
-	 */
-	public FSTObjectOutput getObjectOutput(final byte[] outByte) {
-		FSTObjectOutput fstObjectOutput = getOut();
-		fstObjectOutput.resetForReUse(outByte);
-		return fstObjectOutput;
-	}
 
 	/**
 	 * ignores all serialization related interfaces (Serializable, Externalizable) and serializes all classes using the
@@ -798,7 +663,8 @@ public class FSTConfiguration {
 	 */
 	public Object asObject(final byte b[]) {
 		try {
-			return getObjectInput(b).readObject();
+			final byte[] input = ByteArrayZipper.unzip(b);
+			return getObjectInput(input).readObject();
 		} catch (Exception e) {
 			System.out.println("unable to decode:" + new String(b, 0, 0, Math.min(b.length, 100)));
 			FSTUtil.<RuntimeException> rethrow(e);
@@ -813,27 +679,7 @@ public class FSTConfiguration {
 		FSTObjectOutput objectOutput = getObjectOutput();
 		try {
 			objectOutput.writeObject(object);
-			return objectOutput.getCopyOfWrittenBuffer();
-		} catch (IOException e) {
-			FSTUtil.<RuntimeException> rethrow(e);
-		}
-		return null;
-	}
-
-	/**
-	 * Warning: avoids allocation + copying. The returned byteArray is a direct pointer to underlying buffer. the int
-	 * length[] is expected to have at least on element. The buffer can be larger than written data, therefore length[0]
-	 * will contain written length.
-	 *
-	 * The buffer content must be used (e.g. sent to network, copied to offheap) before doing another asByteArray on the
-	 * current Thread.
-	 */
-	public byte[] asSharedByteArray(final Object object, final int length[]) {
-		FSTObjectOutput objectOutput = getObjectOutput();
-		try {
-			objectOutput.writeObject(object);
-			length[0] = objectOutput.getWritten();
-			return objectOutput.getBuffer();
+			return ByteArrayZipper.zip(objectOutput.getCopyOfWrittenBuffer());
 		} catch (IOException e) {
 			FSTUtil.<RuntimeException> rethrow(e);
 		}

--- a/gama.ui.experiment/src/gama/ui/experiment/menus/SimulationsMenu.java
+++ b/gama.ui.experiment/src/gama/ui/experiment/menus/SimulationsMenu.java
@@ -1,12 +1,12 @@
 /*******************************************************************************************************
  *
- * SimulationsMenu.java, in gama.ui.experiment, is part of the source code of the
- * GAMA modeling and simulation platform (v.2024-06).
+ * SimulationsMenu.java, in gama.ui.experiment, is part of the source code of the GAMA modeling and simulation platform
+ * (v.2024-06).
  *
  * (c) 2007-2024 UMI 209 UMMISCO IRD/SU & Partners (IRIT, MIAT, ESPACE-DEV, CTU)
  *
  * Visit https://github.com/gama-platform/gama for license information and contacts.
- * 
+ *
  ********************************************************************************************************/
 package gama.ui.experiment.menus;
 
@@ -58,8 +58,7 @@ public class SimulationsMenu extends ContributionItem {
 	public static final GamaCommand duplicateCurrentSimulation =
 			new GamaCommand("experiment/experiment.simulations.duplicate", "Duplicate Simulation",
 					"Duplicate the current simulation and add it to the experiment", e -> {
-						byte[] bytes =
-								BinarySerialisation.saveToBytes(GAMA.getRuntimeScope(), GAMA.getSimulation(), true);
+						byte[] bytes = BinarySerialisation.saveToBytes(GAMA.getRuntimeScope(), GAMA.getSimulation());
 						final SimulationAgent sim =
 								GAMA.getExperiment().getAgent().createSimulation(new ParametersSet(), true);
 						GAMA.runAndUpdateAll(() -> {
@@ -101,8 +100,7 @@ public class SimulationsMenu extends ContributionItem {
 				// default -> SerialisationConstants.BINARY_FORMAT;
 				// };
 
-				BinarySerialisation.saveToFile(sim.getScope(), sim, open, ISerialisationConstants.BINARY_FORMAT, true,
-						false);
+				BinarySerialisation.saveToFile(sim.getScope(), sim, open, false);
 			});
 
 	/** The save current simulation and history. */
@@ -118,8 +116,7 @@ public class SimulationsMenu extends ContributionItem {
 						fileSave.setFileName(sim.getModel().getName() + "_" + sim.getCycle(sim.getScope()) + "_cycles"
 								+ ".simulation");
 						String open = fileSave.open();
-						BinarySerialisation.saveToFile(sim.getScope(), sim, open, ISerialisationConstants.BINARY_FORMAT,
-								true, true);
+						BinarySerialisation.saveToFile(sim.getScope(), sim, open, true);
 					});
 
 	/** The replace current simulation. */

